File: hackerrank/problem solving/algorithms/1 - warmup/8 - mini-max sum/MiniMaxSum.java
// https://www.hackerrank.com/challenges/mini-max-sum/problem?isFullScreen=true

import java.util.List;
import java.util.Scanner;
import java.util.Collections;
import java.util.stream.Stream;
import static java.util.stream.Collectors.toList;

public class MiniMaxSum {
    private static Scanner scan;

    public static void main(String[] args) {
        scan = new Scanner(System.in);
        List<Integer> array = readIntArray();
        Collections.sort(array);
        miniMaxSum(array);
    }

        private static List<Integer> readIntArray() {
            List<Integer> array = Stream.of(scan.nextLine().split(" "))
                                  .map(Integer::parseInt)
                                  .collect(toList());
            return array;
        }

        private static void miniMaxSum(List<Integer> array) {
            long totalSum = array.stream().mapToLong(Integer::longValue).reduce(0L, (sum, num) -> sum + num);
            long minSum = totalSum - array.get(array.size() - 1);
            long maxSum = totalSum - array.get(0);
            System.out.println(minSum + " " + maxSum);
        }
}

File: hackerrank/problem solving/algorithms/1 - warmup/8 - mini-max sum/mini_max_sum.cpp
// https://www.hackerrank.com/challenges/mini-max-sum/problem?isFullScreen=true
// C++23

import std;
using namespace std;

vector<long> read_nums(int n);
pair<long, long> calc_min_max_sum(const vector<long>& arr);

int main()
{
    constexpr int n{ 5 };
    vector arr{ read_nums(n) };
    auto [min_sum, max_sum] { calc_min_max_sum(arr) };
    println("{} {}", min_sum, max_sum);
    return 0;
}

vector<long> read_nums(int n)
{
    vector<long> arr(n);
    for (auto& x : arr)
        cin >> x;
    return arr;
}

// n: length of arr
// T: O(n)
// S: O(1) extra space
pair<long, long> calc_min_max_sum(const vector<long>& arr)
{
    long total, min_value, max_value;
    total = min_value = max_value = arr.front();

    for (auto x : arr | views::drop(1)) {
        total += x;
        min_value = min(x, min_value);
        max_value = max(x, max_value);
    }

    return { total - max_value, total - min_value };
}
